### PR TITLE
FIFO and absolute values

### PR DESCRIPTION
**Pull request by mikkeldamsgaard**
Added support for FIFO and reading absolute values even when having the high-pass filter active

### CODE REVIEW

File: src/lis3dh.toit [hunk 1]
```diff
@@ -246,16 +267,22 @@ class Lis3dh:
...
 
+      if fifo: fifo-bits = 0b1100_0000
+
+    reg_.write-u8 FIFO_CTRL_REG_ 0
     reg_.write_u8 CTRL_REG1_ ctrl1
     reg_.write_u8 CTRL_REG2_ ctrl2
     reg_.write_u8 CTRL_REG3_ ctrl3
     reg_.write_u8 CTRL_REG4_ ctrl4
-    reg_.write_u8 CTRL_REG5_ ctrl5
+    reg_.write_u8 CTRL_REG5_ (ctrl5 & 0b1011_1111)
```
Comment (floitsch) [inline src/lis3dh.toit:279]: Add a comment explaining why you have to set ctrl5 first, and then a second time later.

File: src/lis3dh.toit [hunk 2]
```diff
@@ -246,16 +267,22 @@ class Lis3dh:
       else:
         unreachable
 
+      if fifo: fifo-bits = 0b1100_0000
```
Comment (floitsch) [inline src/lis3dh.toit:271]: Add a comment explaining why you have to update the fifo-bits?

File: src/lis3dh.toit [hunk 3]
```diff
@@ -302,6 +329,30 @@ class Lis3dh:
...
+    while true:
+      src := reg_.read-u8 FIFO_SRC_REG_
+      samples = src & 0b001_1111
+      if samples >= min-samples: break
+
+    return List samples : read_acceleration
+
+  /**
+  Resets the reference value to the current measures. This is used to clear any DC component in
+  the measurement data.
```
Comment (floitsch) [inline src/lis3dh.toit]: what's "DC"?
```suggestion
  Resets the reference value to the current measures.
  This is used to clear any DC component in the measurement data.
```

File: src/lis3dh.toit [hunk 4]
```diff
@@ -326,4 +377,14 @@ class Lis3dh:
     the $INTERRUPT_X_HIGH_BIT is set to high if a high event was detected.
   */
   read_interrupt_cause -> int:
-    return reg_.read_u8 INT1_SRC_
+    fifo-ctrl := reg_.read-u8 FIFO_CTRL_REG_
+    ctrl5 := reg_.read-u8 CTRL_REG5_
+
+    if (ctrl5 & 0b1000) != 0 and (fifo-ctrl & 0b1100_0000) == 0b1100_0000:
+      // If latch and fifo
```
Comment (floitsch) [inline src/lis3dh.toit]:
```suggestion
      // If latch and fifo.
```

File: src/lis3dh.toit [hunk 5]
```diff
@@ -326,4 +377,14 @@ class Lis3dh:
...
+      interrupt_source := reg_.read-u8 INT1_SRC_
+      reg_.write-u8 FIFO_CTRL_REG_ 0b0011_1111 & fifo-ctrl
+      reg_.write-u8 FIFO_CTRL_REG_ 0b1100_0000 | fifo-ctrl
```
Comment (floitsch) [inline src/lis3dh.toit:392]: Add a comment what is happening to the fifo.
Is it cleared?

**Review by floitsch — APPROVED** [review]
LGTM.

File: src/lis3dh.toit (hunk 3 above)
Comment (mikkeldamsgaard) [reply src/lis3dh.toit]: DC = Direct Current, as in AC/DC. Electronic Engineers then to call everything that is not oscillating DC

File: src/lis3dh.toit [hunk 6]
```diff
@@ -267,20 +267,22 @@ class Lis3dh:
...
 
+      // Set up the fifo in Stream-to-FIFO mode.
       if fifo: fifo-bits = 0b1100_0000
 
     reg_.write-u8 FIFO_CTRL_REG_ 0
     reg_.write_u8 CTRL_REG1_ ctrl1
     reg_.write_u8 CTRL_REG2_ ctrl2
     reg_.write_u8 CTRL_REG3_ ctrl3
     reg_.write_u8 CTRL_REG4_ ctrl4
+    // According to the datasheet, the FIFO enable bit should be delayed to after interupts.
```
Comment (floitsch) [inline src/lis3dh.toit]:
```suggestion
    // According to the datasheet, the FIFO enable bit should be delayed to after interrupts.
```

File: src/lis3dh.toit [hunk 7]
```diff
@@ -267,20 +267,22 @@ class Lis3dh:
...
     reg_.write_u8 CTRL_REG5_ (ctrl5 & 0b1011_1111)
 
     if needs_reference: reset-reference
     reg_.write_u8 INT1_THS_ int1_threshold
     reg_.write_u8 INT1_DURATION_ int1_duration
     reg_.write_u8 INT1_CFG_ int1_cfg
-    reg_.write_u8 CTRL_REG5_ ctrl5
+    reg_.write_u8 CTRL_REG5_ ctrl5 // Set the FIF enable bit now.
```
Comment (floitsch) [inline src/lis3dh.toit]:
```suggestion
    reg_.write_u8 CTRL_REG5_ ctrl5 // Set the FIFO enable bit now.
```

**Review by floitsch — APPROVED** [review]
LGTM.

**Comment by floitsch** [comment]
Thanks.

**Pull request merged**